pkg/server/apiextensions.go

```diff
@@ -31,137 +31,232 @@ import (
-	// before we can try to get said ClusterWorkspace, but if we fail listing because the ClusterWorkspace doesn't
-	// exist, we'll never be able to create it. Only check if the target workspace exists for
-	// non-default keys.
-	if cluster.Name != helper.RootCluster && c.workspaceLister != nil {
-		targetWorkspaceKey := helper.WorkspaceKey(orgName, ws)
-		workspace, err := c.workspaceLister.Get(targetWorkspaceKey)
-		if err != nil && !apierrors.IsNotFound(err) {
-			// Only return errors other than not-found. If we couldn't find the workspace, let's continue
-			// to list the CRDs in ctx's logical cluster, at least until we have proper ClusterWorkspace permissions
-			// requirements in place (i.e. reject all requests to a logical cluster if there isn't a
-			// ClusterWorkspace for it). Otherwise, because this method is used for API discovery, you'd have
-			// a weird situation where you could create a CRD but not be able to perform CRUD operations
-			// on its CRs with kubectl (because it relies on discovery, and returning [] when we can't
-			// find the ClusterWorkspace would mean CRDs from this logical cluster wouldn't be in discovery).
-			return nil, err
+	for _, apiBinding := range apiBindings {
+		if apiBinding.ClusterName != cluster.Name {
+			continue
+		}
+		if apiBinding.Status.Phase != apisv1alpha1.APIBindingPhaseBound {
+			// TODO(ncdc): what if it's in the middle of rebinding and it takes some time - we won't include any
+			// CRDs that were previously bound...
+			continue
 		}
 
-		if workspace != nil && workspace.Spec.InheritFrom != "" {
-			if workspace.Spec.InheritFrom == helper.RootCluster {
-				// HACK: allow inheriting from the OrganizationCluster logical cluster
-				inheriting = true
-				inheritFrom = helper.RootCluster
-			} else {
-				// Make sure the source workspace exists
-				wsName := workspace.Spec.InheritFrom
-				orgClusterName := orgName
-				if strings.ContainsRune(workspace.Spec.InheritFrom, ':') {
-					orgClusterName, err = helper.ParentClusterName(workspace.Spec.InheritFrom)
-					if err != nil {
-						return nil, err
-					}
-					orgName, wsName, err = helper.ParseLogicalClusterName(workspace.Spec.InheritFrom)
-					if err != nil {
-						return nil, err
-					}
-				}
-				sourceWorkspaceKey := helper.WorkspaceKey(orgClusterName, wsName)
-				_, err := c.workspaceLister.Get(sourceWorkspaceKey)
-				switch {
-				case err == nil:
-					inheriting = true
-					inheritFrom = helper.EncodeOrganizationAndClusterWorkspace(orgName, wsName)
-				case apierrors.IsNotFound(err):
-					// A NotFound error is ok. It means we can't inherit but we should still proceed below to list.
-				default:
-					// Only error if there was a problem checking for workspace existence
-					return nil, err
-				}
+		for _, boundResource := range apiBinding.Status.BoundResources {
+			crdKey := clusters.ToClusterAwareKey("system:bound-crds", boundResource.Schema.UID)
```

Comment:
(No action required) Maybe this special workspace name can be defined as constant?

Reply:
done